src/lib/comms/sol-lwm2m-common.h

```diff
@@ -0,0 +1,679 @@
+            SOL_LWM2M_TLV_API_VERSION)) { \
+            SOL_WRN("Couldn't handle tlv that has unsupported version " \
+                "'%u', expected version is '%u'", \
+                (_tlv)->api_version, SOL_LWM2M_TLV_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0);
+#define LWM2M_RESOURCE_CHECK_API(_resource, ...) \
+    do { \
+        if (SOL_UNLIKELY((_resource)->api_version != \
+            SOL_LWM2M_RESOURCE_API_VERSION)) { \
+            SOL_WRN("Couldn't handle resource that has unsupported version " \
+                "'%u', expected version is '%u'", \
+                (_resource)->api_version, SOL_LWM2M_RESOURCE_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0);
+#define LWM2M_RESOURCE_CHECK_API_GOTO(_resource, _label) \
+    do { \
+        if (SOL_UNLIKELY((_resource).api_version != \
+            SOL_LWM2M_RESOURCE_API_VERSION)) { \
+            SOL_WRN("Couldn't handle resource that has unsupported version " \
+                "'%u', expected version is '%u'", \
+                (_resource).api_version, SOL_LWM2M_RESOURCE_API_VERSION); \
+            goto _label; \
+        } \
+    } while (0);
+#define LWM2M_OBJECT_CHECK_API(_obj, ...) \
+    do { \
+        if (SOL_UNLIKELY((_obj).api_version != \
+            SOL_LWM2M_OBJECT_API_VERSION)) { \
+            SOL_WRN("Couldn't handle object that has unsupported version " \
+                "'%u', expected version is '%u'", \
+                (_obj).api_version, SOL_LWM2M_OBJECT_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0);
+#define LWM2M_OBJECT_CHECK_API_GOTO(_obj, _label) \
+    do { \
+        if (SOL_UNLIKELY((_obj).api_version != \
+            SOL_LWM2M_OBJECT_API_VERSION)) { \
+            SOL_WRN("Couldn't handle object that has unsupported version " \
+                "'%u', expected version is '%u'", \
+                (_obj).api_version, SOL_LWM2M_OBJECT_API_VERSION); \
+            goto _label; \
+        } \
+    } while (0);
+#else
+#define LWM2M_TLV_CHECK_API(_tlv, ...)
+#define LWM2M_RESOURCE_CHECK_API(_resource, ...)
+#define LWM2M_OBJECT_CHECK_API(_obj, ...)
+#define LWM2M_RESOURCE_CHECK_API_GOTO(_resource, _label)
+#define LWM2M_OBJECT_CHECK_API_GOTO(_obj, _label)
+#endif
+
+enum tlv_length_size_type {
+    LENGTH_SIZE_CHECK_NEXT_TWO_BITS = 0,
+    LENGTH_SIZE_8_BITS = 8,
+    LENGTH_SIZE_16_BITS = 16,
+    LENGTH_SIZE_24_BITS = 24
+};
+
+enum lwm2m_parser_args_state {
+    STATE_NEEDS_DIGIT = 0,
+    STATE_NEEDS_COMMA_OR_EQUAL = (1 << 1),
+    STATE_NEEDS_COMMA = (1 << 2),
+    STATE_NEEDS_APOSTROPHE = (1 << 3),
+    STATE_NEEDS_CHAR_OR_DIGIT = (1 << 4),
+};
+
+struct lifetime_ctx {
+    struct sol_timeout *timeout;
+    uint32_t lifetime;
+};
+
+struct sol_lwm2m_client_object {
+    struct sol_ptr_vector instances;
+    uint16_t id;
+};
+
+static void
+send_ack_if_needed(struct sol_coap_server *coap, struct sol_coap_packet *msg,
```

Comment:
I don't think we should have all this functions in the header file. Move them to sol-lwm2m-common.c